### PR TITLE
Arm warrior spec implementation

### PR DESCRIPTION
**Pull request by matardarix**
Hi, i'm playing as an Arm Warrior since the end of WoD and i'm using WoWAnalyzer since Legion. Since the arm spec does not have a maintener i wanted to give it a try. Please tell me if i should change anything before this analysis could be merge. 

**What's New:** 
- Checklist: Core abilities and offensive cooldowns, defensive cooldowns, downtime and well prepared.
image
- Talents statistics: 
Tier 15: War Machine uptime / Execute damage under Sudden Death / Skullspitter average damage
Tier 30: Impending Victory heal / Storm Bolt stun
Tier 60: Second Wind heal
Tier 75: Cleave average damage / Warbreaker average damage (Warbreaker + 30% increase damage during debuff)
Tier 90: Avatar increased damage
Tier 100: Anger Management (moved from statistic) / Ravager damage

image
image

- Azerite Traits: Executioner's Precision: Suggestion on wasted stack, Seismic Wave: DPS, Test of Might: Strength buff average

image
image
image

- Mortal Strike on Execute range without Executioner's Precision:
image

**What's change**

- Statistics
image

- Defensive Stance: Suggestion when you reduced your own damage more than incoming damage. Suggestion when you never used this talent. (I'm still working on improving this) 

- Anger Managment: Correction applied on reduced cooldowns.

I plan to continue working on this spec but i would prefer that these changes be validated by both the community and the mainteners.
Again please let me know if i should add or change anything, i'll try to do it as soon as i can.

### CODE REVIEW

File: src/parser/warrior/arms/CONFIG.js [hunk 1]
```diff
@@ -25,6 +25,10 @@ export default {
         This spec needs a focused maintainer. If you want to give it a try, check <a href="https://github.com/WoWAnalyzer/WoWAnalyzer">GitHub</a> for more information.
       </Warning><br />
 
+      If you find any issues or have something you'd like to see added, open an issue on <a href="https://github.com/WoWAnalyzer/WoWAnalyzer/issues/new">GitHub</a>, contact us on <a href="https://discord.gg/AxphPxU">Discord</a> or DM us on Discord.<br /><br />
+
+      Make sure to check out the <a href="https://discordapp.com/invite/Skyhold">Warrior Class Discord</a> if you need more specific advice or a more detailed guide than the ones available on <a href="https://www.icy-veins.com/wow/arms-warrior-pve-dps-guide">Icy-Veins</a> and <a href="https://www.wowhead.com/arms-warrior-guide">wowhead</a>.<br /><br />
+
       <Warning>
         Hi <kbd>@Zerotorescue</kbd> here. I updated this spec to be compatible with BFA, but I'm no Arms expert. If there's anything wrong please @poke me on Discord and I'll fix it if I can.
       </Warning>
```
Comment (Pewtro) [inline src/parser/warrior/arms/CONFIG.js]: You should probably cleanup this config so as to remove line 32->34 and 22->26 since it is now supported. If line 20 isn't relevant, change that aswell.

File: src/parser/warrior/arms/modules/Abilities.js [hunk 2]
```diff
@@ -75,7 +83,7 @@ class Abilities extends CoreAbilities {
         },
         castEfficiency: {
           suggestion: true,
-          recommendedEfficiency: 0.6, // I don't know at what efficiency this talent becomes worth it so I'm keeping it save
+          recommendedEfficiency: 0.9, // I don't know at what efficiency this talent becomes worth it so I'm keeping it save
         },
         enabled: combatant.hasTalent(SPELLS.WARBREAKER_TALENT.id), // replaces Colussus Smash
```
Comment (Pewtro) [inline src/parser/warrior/arms/modules/Abilities.js]: Is that comment still relevant since you upped cast efficiency drastically?

File: src/parser/warrior/arms/modules/Abilities.js [hunk 3]
```diff
@@ -177,14 +193,6 @@ class Abilities extends CoreAbilities {
...
         gcd: null,
-        castEfficiency: {
-          // TODO: Add a statistic to show how much damage was lost due to defensive stance
-          // TODO: Add a suggestion regarding having this up too much
-          // TODO: Add a suggestion regarding having this up too little, and then deactivate the cast efficiency since uptime is more accurate
-          suggestion: true,
-          recommendedEfficiency: 0.4,
-          extraSuggestion: 'If you\'re using this talent over something that passively increases your mobility or survivability, you better use it.',
-        },
         enabled: combatant.hasTalent(SPELLS.DEFENSIVE_STANCE_TALENT.id),
```
Comment (Pewtro) [inline src/parser/warrior/arms/modules/Abilities.js]: Are the todos not relevant anymore if you've chosen the talent? If not, it's fine to remove them. If they are relevant, you might want to open an issue for them.

File: src/parser/warrior/arms/modules/core/Dots/RendUptime.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import StatisticListBoxItem from 'interface/others/StatisticListBoxItem';
+
+import ExecuteRange from '../../core/Execute/ExecuteRange';
+
+class RendUptime extends Analyzer {
+  static dependencies = {
+    enemies: Enemies,
+    executeRange: ExecuteRange,
+  };
+
```
Comment (Pewtro) [inline src/parser/warrior/arms/modules/core/Dots/RendUptime.js]: executeRange is never used

File: src/parser/warrior/arms/modules/core/Dots/index.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+  statistic() {
+    return (
+      <StatisticsListBox
+        position={STATISTIC_ORDER.CORE(3)}
+        title="DoT uptimes">
+        {this.deepwoundsUptime.subStatistic()}
+        {this.selectedCombatant.hasTalent(SPELLS.REND_TALENT.id) ? this.rendUptime.subStatistic() : null}
+      </StatisticsListBox>
+    );
```
Comment (Pewtro) [inline src/parser/warrior/arms/modules/core/Dots/index.js]: You can skip the ternary by using
```JS
{this.rendUptime.active && this.rendUptime.subStatistic()}
```

File: src/parser/warrior/arms/modules/core/Execute/ExecuteRange.js [hunk 6]
```diff
@@ -1,23 +1,24 @@
...
+
   enemyMap = {};
 
   on_byPlayer_damage(event) {
-    if(event.targetIsFriendly) {
-      return;
-    }
-
+    if(event.targetIsFriendly) return;
     const targetString = encodeTargetString(event.targetID, event.targetInstance);
```
Comment (Pewtro) [inline src/parser/warrior/arms/modules/core/Execute/ExecuteRange.js]: fwiw we generally prefer using the version that was there before for consistency across the project

File: src/parser/warrior/arms/modules/features/AlwaysBeCasting.js [hunk 7]
```diff
@@ -18,8 +27,6 @@ class AlwaysBeCasting extends CoreAlwaysBeCasting {
           .regular(recommended + 0.15).major(recommended + 0.2);
       });
```
Comment (Pewtro) [inline src/parser/warrior/arms/modules/features/AlwaysBeCasting.js]: you can remove the .regular(....) line.

File: src/parser/warrior/arms/modules/spells/azeritetraits/ExecutionersPrecision.js [hunk 8]
```diff
@@ -7,30 +7,36 @@ import SPELLS from 'common/SPELLS';
...
+  }
+
+  procs = 0;
+  wastedProcs = 0;
 
   on_byPlayer_cast(event) {
-    if (SPELLS.EXECUTE.id !== event.ability.guid) {
-      return;
-    }
+    if (SPELLS.EXECUTE.id !== event.ability.guid) return;
```
Comment (Pewtro) [inline src/parser/warrior/arms/modules/spells/azeritetraits/ExecutionersPrecision.js]: Same as earlier, we prefer using the approach that was there before for consistency.

File: src/parser/warrior/arms/modules/spells/azeritetraits/ExecutionersPrecision.js [hunk 9]
```diff
@@ -7,30 +7,36 @@ import SPELLS from 'common/SPELLS';
...
 
-    this.executes += 1;
+    this.procs += 1;
     const enemy = this.enemies.getEntity(event);
-    if (!enemy) {
-      return;
-    }
-    const executionersPrecision = enemy.getBuff(SPELLS.EXECUTIONERS_PRECISION.id);
+    if (!enemy) return;
+
```
Comment (Pewtro) [inline src/parser/warrior/arms/modules/spells/azeritetraits/ExecutionersPrecision.js]: Same as earlier, we prefer using the approach that was there before for consistency.

File: src/parser/warrior/arms/modules/spells/azeritetraits/SeismicWave.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  damage = 0;
+
+  constructor(...args) {
+    super(...args);
+    this.active = this.selectedCombatant.hasTrait(SPELLS.SEISMIC_WAVE.id);
+  }
+
+  on_byPlayer_damage(event) {
+    if (event.ability.guid !== SPELLS.SEISMIC_WAVE_DAMAGES.id) return;
+    this.damage += (event.amount || 0) + (event.absorbed || 0);
```
Comment (Pewtro) [inline src/parser/warrior/arms/modules/spells/azeritetraits/SeismicWave.js]:
```JS
if (event.ability.guid !== SPELLS.SEISMIC_WAVE_DAMAGES.id) {
 return;
}
```
Is preferred

**Comment by matardarix** [comment]
Thanks for your advice, be sure it will be taken into account for the future. Please take anytime you need to review this.

File: src/parser/warrior/arms/modules/Abilities.js (hunk 2 above)
Comment (matardarix) [reply src/parser/warrior/arms/modules/Abilities.js]: I removed this comment since it is now irrelevant

File: src/parser/warrior/arms/modules/Abilities.js (hunk 3 above)
Comment (matardarix) [reply src/parser/warrior/arms/modules/Abilities.js]: The todos have been moved on the Defensive Stance file ('src\parser\warrior\arms\modules\talents\DefensiveStance.js), some have been already added in this file

**Comment by Sharrq** [comment]
Oh, you might also want to include some changelog entries for these updates/changes.

File: src/parser/warrior/arms/modules/spells/azeritetraits/TestOfMight.js [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+    proc = 0;
+    strengthOnToM = 0;
+    totalStrength = 0;
+    rageSpendDuringDebuff = 0;
+
+    constructor(...args) {
+        super(...args);
+        this.active = this.selectedCombatant.hasTrait(SPELLS.TEST_OF_MIGHT.id);
+        if(!this.active) {
```
[the remark this answers is not in the document]
Comment (matardarix) [reply src/parser/warrior/arms/modules/spells/azeritetraits/TestOfMight.js]: I actually can't get rid of this one because of the variable initialization underneath, otherwise i get an error: "TypeError: can't convert undefined to object".

[the remark this answers is not in the document]
Comment (Sharrq) [reply src/parser/warrior/arms/modules/spells/azeritetraits/TestOfMight.js]: Fair enough

File: src/parser/warrior/arms/modules/talents/DefensiveStance.js [hunk 12]
```diff
@@ -98,6 +104,21 @@ class DefensiveStance extends Analyzer {
       />
     );
   }
+  
+  suggestions(when) {
+    when(this.totalDamageLost).isGreaterThan(this.totalDamageMitigated)
+      .addSuggestion((suggest, dl, dr) => {
+        return suggest('The damage you inflicted were more reduced than those you received. Try to reduce the uptime of Defensive Stance when you take less damage.')
```
Comment (Sharrq) [inline src/parser/warrior/arms/modules/talents/DefensiveStance.js]: Still not sure what this means ...

Are you trying to say they had more damage taken than damage dealt? As in the boss did more damage to them than they did to the boss? And that they should save Defensive Stance for when they are about to take damage?

Or are you trying to say they blocked more damage than they took? As in Defensive stance prevented more damage than damage that actually hit them?

I am also not quite sure what Damage Lost is

Comment (matardarix) [reply src/parser/warrior/arms/modules/talents/DefensiveStance.js]: So the totalDamageLost stand for the 10% decreased damage dealt by the player along the fight and the totalDamageMitigated for the 20% decreased damage taken by the player while he had the Defensive Stance active. As Defensive Stance could be activated and desactivated (after the 6sc cooldown) this could means that the Defensive Stance was active for too long during the fight

Comment (Sharrq) [reply src/parser/warrior/arms/modules/talents/DefensiveStance.js]: oh does Defensive Stance reduce the damage you deal?

Comment (matardarix) [reply src/parser/warrior/arms/modules/talents/DefensiveStance.js]: Yes it's reduces the damage you take by 20% but you lost 10% of your damages while it's active :)

Comment (Sharrq) [reply src/parser/warrior/arms/modules/talents/DefensiveStance.js]: Oh, i didnt realize Defensive Stance reduced the damage you deal also. 

Maybe something like this would be better then?
"While Defensive Stance was up, your damage done was reduced by more than the damage you mitigated. Ensure that you are only using Defensive Stance when you are about to take a lot of damage and that you cancel it quickly to minimize the time spent dealing less damage."

or some variation of that if its not entirely accurate.

Comment (Sharrq) [reply src/parser/warrior/arms/modules/talents/DefensiveStance.js]: No problem at all, just want to make sure that people coming to the site understand it :)

**Comment by matardarix** [comment]
Alright, thanks for your reviews and corrections, and again, sorry for the size of this PR, i'll do smaller next time.

File: src/parser/warrior/arms/modules/features/RageUsage.js [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+    SPELLS.WHIRLWIND.id,
+    SPELLS.HAMSTRING.id,
+    SPELLS.MORTAL_STRIKE.id,
+    SPELLS.EXECUTE.id,
+    SPELLS.SLAM.id,
+    SPELLS.IMPENDING_VICTORY_TALENT.id,
+    SPELLS.REND_TALENT.id,
+];
+
+class RageUsage extends Analyzer {
```
Comment (sMteX) [inline src/parser/warrior/arms/modules/features/RageUsage.js]: I'm not sure what the *general consensus* on doughnut charts are, some like it, some hate it (\*cough\* emallson \*cough\*), and some like me are kinda neutral. I think they are visually pleasing, but the fact that they show pretty much the same thing as the Resource tab (if implemented) and the fact that they can become very cluttered if many spells are displayed, doesn't really tip the scales in their favor. Related discussion in #2333 .

File: src/parser/warrior/arms/modules/spells/azeritetraits/TestOfMight.js [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+        
+        this.strengthOnToM += this.strength * (this.rageSpendDuringDebuff / RAGE_PER_PROC);
+        this.totalStrength += this.strengthOnToM;
+
+        this.rageSpendDuringDebuff = 0;
+        this.strengthOnToM = 0;
+    }
+
+    on_byPlayer_cast(event) {
+        if (!event.classResources ||
```
Comment (sMteX) [inline src/parser/warrior/arms/modules/spells/azeritetraits/TestOfMight.js]: Maybe this could be simplified to something like (I haven't checked, not 100% sure)
```javascript
const rage = event.classResources.find(e => e.type === RESOURCE_TYPES.RAGE.id);
if (!rage) { 
    return;
}
....
    this.rageSpendDuringDebuff += rage.cost / 10;
```
Cast events *should* always have a `classResources` field, and if `find()` can't find anything, it returns `undefined`.

File: src/parser/warrior/arms/modules/spells/azeritetraits/TestOfMight.js [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+        return (this.totalStrength / this.proc * this.uptime).toFixed(0);
+    }
+
+    statistic() {
+        return (
+          <TraitStatisticBox
+            position={STATISTIC_ORDER.OPTIONAL()}
+            trait={SPELLS.TEST_OF_MIGHT.id}
+            value={(
+              <>
```
Comment (sMteX) [inline src/parser/warrior/arms/modules/spells/azeritetraits/TestOfMight.js]: This should probably be just 
```jsx
value={`${this.averageStrength} average Strength`}
```
not sure about the role of the newline at the end.

File: src/parser/warrior/arms/modules/core/Overpower.js [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+    wastedProc = 0;
+
+    on_byPlayer_cast(event) {
+        if (event.ability.guid !== SPELLS.OVERPOWER.id) {
+            return;
+        }
+
+        this.proc += 1;
+        const overpower = this.selectedCombatant.getBuff(SPELLS.OVERPOWER.id);
+        if (!this.executeRange.isTargetInExecuteRange(event) && !this.selectedCombatant.hasTrait(SPELLS.EXECUTIONERS_PRECISION_TRAIT.id)) {
```
Comment (sMteX) [inline src/parser/warrior/arms/modules/core/Overpower.js]: Not incorrect, although if you want to, you can reduce the verbosity a little by setting a constant in constructor, like 
```javascript
this.hasEP = this.selectedCombatant.hasTrait(SPELLS.EXECUTIONERS_PRECISION.id)
```

File: src/parser/warrior/arms/modules/core/Overpower.js [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+            if (overpower !== undefined && overpower.stacks === 2 && this.spellUsable.isAvailable(SPELLS.MORTAL_STRIKE.id)) {
```
Comment (sMteX) [inline src/parser/warrior/arms/modules/core/Overpower.js]: Checking just for `overpower && overpower.stacks === 2 && this.spellUsable.isAvailable(SPELLS.MORTAL_STRIKE.id)` should be enough, but again, it's not incorrect to explicitly check for `undefined`.

File: src/parser/warrior/arms/modules/talents/DefensiveStance.js [hunk 18]
```diff
@@ -39,9 +47,7 @@ class DefensiveStance extends Analyzer {
   constructor(...args) {
     super(...args);
     this.active = this.selectedCombatant.hasTalent(SPELLS.DEFENSIVE_STANCE_TALENT.id);
-    if (!this.active) {
-      return;
-    }
+
     this.addEventListener('damage', this.handleDamageTaken, {
```
[the remark this answers is not in the document]
Comment (matardarix) [reply src/parser/warrior/arms/modules/talents/DefensiveStance.js]: This part of the code isn't mine, i've just added suggestions on it, that's why, but i'll change it for consistency

File: src/parser/warrior/arms/modules/features/RageUsage.js (hunk 13 above)
Comment (matardarix) [reply src/parser/warrior/arms/modules/features/RageUsage.js]: I could get rid of this part and implement the Ressource tab later on if you prefer

Comment (sMteX) [reply src/parser/warrior/arms/modules/features/RageUsage.js]: Unless someone else reviews this and decides you should delete it, I'd say keep it 😄 Just pointing the facts out, that it's pretty much the same information.

File: src/parser/warrior/arms/modules/core/Overpower.js (hunk 17 above)
Comment (matardarix) [reply src/parser/warrior/arms/modules/core/Overpower.js]: If i don't check it for undefined i get a TypeError 
image

File: src/parser/warrior/arms/modules/features/RageUsage.js (hunk 13 above)
Comment (matardarix) [reply src/parser/warrior/arms/modules/features/RageUsage.js]: Alright, i might change this in the futur anyway to avoid having too much informations not that meaningfull on the statistic part but i leave it that way for now

File: src/parser/warrior/arms/modules/core/Overpower.js (hunk 17 above)
Comment (sMteX) [reply src/parser/warrior/arms/modules/core/Overpower.js]: That's peculiar (seeing as `executionersPrecision` is not even mentioned here), but we can look into that later.

Comment (matardarix) [reply src/parser/warrior/arms/modules/core/Overpower.js]: Well... that's probably because i was refering to the wrong file, for the Executioner's Precision trait (in /modules/spells/azeritetraits/) where i also have an undefined check. Sorry it looks like i got lost at some point :/

Comment (sMteX) [reply src/parser/warrior/arms/modules/core/Overpower.js]: The undefined check is still probably better to leave in, just as a shorthand - `overpower && overpower.stacks ....`, so it doesn't crash if `overpower` is null or undefined (trying to access `stacks` of null or undefined).

Comment (sMteX) [reply src/parser/warrior/arms/modules/core/Overpower.js]: Can you fix this line as well (few lines above the one you fixed)? It's the same thing, maybe you just forgot 😅 Otherwise good from my side. I asked if someone more experienced could take one last look at entire PR if we haven't missed anything important so we may finally merge it in 😃

Comment (matardarix) [reply src/parser/warrior/arms/modules/core/Overpower.js]: Yeah sorry missed that one 😅 it's done. Thanks for your reviews !

File: src/parser/warrior/arms/CombatLogParser.js [hunk 19]
```diff
@@ -1,21 +1,49 @@
 import CoreCombatLogParser from 'parser/core/CombatLogParser';
 import DamageDone from 'parser/shared/modules/DamageDone';
 
+// Features
+import Checklist from './modules/features/Checklist/Module';
 import Abilities from './modules/Abilities';
 import AlwaysBeCasting from './modules/features/AlwaysBeCasting';
-import BattleCryDamage from './modules/core/BattleCryDamage';
-import BattleCryGCD from './modules/core/BattleCryGCD';
 import CooldownThroughputTracker from './modules/features/CooldownThroughputTracker';
```
Comment (emallson) [inline src/parser/warrior/arms/CombatLogParser.js]: it doesn't look like arms has any cooldown tracking implemented. this should be cut out for now to avoid an empty cooldowns tab (see image)
screenshot from 2018-11-04 19-03-03

File: src/parser/warrior/arms/modules/talents/AngerManagement.js [hunk 20]
```diff
@@ -55,22 +61,20 @@ class AngerManagement extends Analyzer {
...
     }, '');
   }
 
-  statistic() {
+  subStatistic() {
     return (
-      <StatisticBox
-        icon={<SpellIcon id={SPELLS.ANGER_MANAGEMENT_TALENT.id} />}
+      <StatisticListBoxItem
+        title={<><SpellLink id={SPELLS.ANGER_MANAGEMENT_TALENT.id} /> cooldown reduction</>}
```
Comment (emallson) [inline src/parser/warrior/arms/modules/talents/AngerManagement.js]: this is running into the value for me. you should abbreviate "cooldown reduction" somehow (I'd use CDR, personally. the tooltip clarifies what the time means)
screenshot from 2018-11-04 19-05-14

File: src/parser/warrior/arms/CHANGELOG.js [hunk 21]
```diff
@@ -1,10 +1,16 @@
 import React from 'react';
 
-import { Aelexe, Zerotorescue, Sharrq } from 'CONTRIBUTORS';
+
+import { Aelexe, Zerotorescue, Sharrq, Matardarix } from 'CONTRIBUTORS';
 import SPELLS from 'common/SPELLS';
 import SpellLink from 'common/SpellLink';
 
 export default [
+  {
```
Comment (emallson) [inline src/parser/warrior/arms/CHANGELOG.js]: you might consider using a `<ul>` here to make it easier to read. not necessary, just a thought

File: src/parser/warrior/arms/modules/core/Overpower.js [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+
+    on_byPlayer_cast(event) {
+        if (event.ability.guid !== SPELLS.OVERPOWER.id) {
+            return;
+        }
+
+        this.proc += 1;
+        const overpower = this.selectedCombatant.getBuff(SPELLS.OVERPOWER.id);
+        if (!this.executeRange.isTargetInExecuteRange(event) && !this.hasEP) {
+            if (overpower && overpower.stacks === 2 && this.spellUsable.isAvailable(SPELLS.MORTAL_STRIKE.id)) {
```
Comment (emallson) [inline src/parser/warrior/arms/modules/core/Overpower.js]: we generally try to avoid nesting ifs because it can make the branching hard to follow. you can restructure this nested-if into an early return (using the inner condition, which is shared across both branches) and then an if/elseif

File: src/parser/warrior/arms/modules/features/RageUsage.js (hunk 13 above)
Comment (emallson) [reply src/parser/warrior/arms/modules/features/RageUsage.js]: as much as i hate doughnut charts, I'm not gonna hold this PR up for including one :P. when you get around to implementing the Resources tab, this can get the axe

File: src/parser/warrior/arms/CombatLogParser.js (hunk 19 above)
Comment (Gebuz) [reply src/parser/warrior/arms/CombatLogParser.js]: #2589 adds cooldown tracking for debuffs, so once that is merged you could add Colossus Smash and War Breaker.

Comment (matardarix) [reply src/parser/warrior/arms/CombatLogParser.js]: Alright, i'll add that in a futur PR, thank's ! I've put the Cooldowns Tab on comment for now

File: src/parser/warrior/arms/CHANGELOG.js (hunk 21 above)
Comment (matardarix) [reply src/parser/warrior/arms/CHANGELOG.js]: Looks like this now
image
